File: data/src/main/java/ch/srg/dataProvider/integrationlayer/data/ImageUrl.java
package ch.srg.dataProvider.integrationlayer.data;

import androidx.annotation.NonNull;

import java.io.Serializable;
import java.util.Objects;

/**
 * Copyright (c) SRG SSR. All rights reserved.
 * <p>
 * License information is available from the LICENSE file.
 */
public final class ImageUrl implements Serializable {
    @NonNull
    private final String url;

    public ImageUrl(@NonNull String url) {
        this.url = url;
    }

    public IlImage getIlImage() {
        return new IlImage(url);
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        ImageUrl imageUrl = (ImageUrl) o;
        return url.equals(imageUrl.url);
    }

    /**
     * Only for internal use!
     *
     * @return the undecorated url
     */
    @NonNull
    public String getRawUrl() {
        return url;
    }

    @Override
    public int hashCode() {
        return Objects.hash(url);
    }

    @Override
    public String toString() {
        return url;
    }
}

File: data/src/main/java/ch/srg/dataProvider/integrationlayer/data/ImageUrl.kt
/*
 * Copyright (c) SRG SSR. All rights reserved.
 * License information is available from the LICENSE file.
 */
@file:Suppress("MemberVisibilityCanBePrivate")

package ch.srg.dataProvider.integrationlayer.data

import ch.srg.dataProvider.integrationlayer.data.serializer.ImageUrlSerializer
import kotlinx.serialization.Serializable

/**
 * Image url
 *
 * @property rawUrl Internal image url, to retrieve the url use [ImageUrl.decorated].
 */
@Serializable(with = ImageUrlSerializer::class)
data class ImageUrl(
    /**
     * @return the undecorated url
     */
    val rawUrl: String
) {

    /**
     * Decorated
     *
     * @param decorator The [ImageUrlDecorator] used to decorate the [rawUrl].
     * @param widthPixels The width of the image.
     * @return The decorated [rawUrl].
     */
    fun decorated(decorator: ImageUrlDecorator, widthPixels: Int): String {
        return decorator.decorate(rawUrl, widthPixels)
    }

    @Deprecated("Using toString is not recommended in this case.", replaceWith = ReplaceWith("rawUrl"))
    override fun toString(): String {
        return rawUrl
    }
}
